crates/broker/src/db/mod.rs

```diff
@@ -23,47 +24,83 @@ mod fuzz_db;
 
 #[derive(Error, Debug)]
 pub enum DbError {
-    #[error("Order key {0} not found in DB")]
+    #[error("{code} Order key {0} not found in DB", code = self.code())]
     OrderNotFound(String),
 
-    #[error("Batch key {0} not found in DB")]
+    #[error("{code} Batch key {0} not found in DB", code = self.code())]
     BatchNotFound(usize),
 
-    #[error("Batch key {0} has no aggreagtion state")]
+    #[error("{code} Batch key {0} has no aggreagtion state", code = self.code())]
     BatchAggregationStateIsNone(usize),
 
     #[cfg(test)]
-    #[error("Batch insert failed {0}")]
+    #[error("{code} Batch insert failed {0}", code = self.code())]
     BatchInsertFailure(usize),
 
-    #[error("DB Missing column value: {0}")]
+    #[error("{code} DB Missing column value: {0}", code = self.code())]
     MissingElm(&'static str),
 
-    #[error("SQL error")]
+    #[error("{code} SQL error {0}", code = self.code())]
     SqlErr(#[from] sqlx::Error),
 
-    #[error("SQL Migration error")]
+    #[error("{code} SQL Pool timed out {0}", code = self.code())]
+    SqlPoolTimedOut(sqlx::Error),
+
+    #[error("{code} SQL Database locked {0}", code = self.code())]
+    SqlDatabaseLocked(anyhow::Error),
+
+    #[error("{code} SQL Migration error", code = self.code())]
     MigrateErr(#[from] sqlx::migrate::MigrateError),
 
-    #[error("JSON serialization err")]
+    #[error("{code} JSON serialization error", code = self.code())]
     JsonErr(#[from] serde_json::Error),
 
-    #[error("Invalid order id")]
+    #[error("{code} Invalid order id", code = self.code())]
     InvalidOrderId(#[from] RuintParseErr),
 
-    #[error("Invalid block number: {0}")]
+    #[error("{code} Invalid block number: {0}", code = self.code())]
     BadBlockNumb(String),
 
-    #[error("Failed to set last block")]
+    #[error("{code} Failed to set last block", code = self.code())]
     SetBlockFail,
 
-    #[error("Invalid order id: {0} missing field: {1}")]
+    #[error("{code} Invalid order id: {0} missing field: {1}", code = self.code())]
     InvalidOrder(String, &'static str),
 
-    #[error("Invalid max connection env var value")]
+    #[error("{code} Invalid max connection env var value", code = self.code())]
     MaxConnEnvVar(#[from] std::num::ParseIntError),
 }
 
+impl CodedError for DbError {
+    fn code(&self) -> &str {
+        match self {
+            DbError::SqlDatabaseLocked(_) => "[B-DB-001]",
+            DbError::SqlPoolTimedOut(_) => "[B-DB-002]",
+            _ => "[B-DB-500]",
+        }
+    }
+}
+trait SqlxResultMapper<T> {
+    fn map_sqlx_err(self) -> Result<T, DbError>;
+}
```

Comment:
Another approach is to define `From<sqlx::Error> for DbError` and use `result?` or `result.map_err(Into::into)` if its in the return position.
https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=b7ccc2713ebb05168515011e32d9cded